versioned/storage/cleanup/src/test/java/org/projectnessie/versioned/storage/cleanup/TestCutHistory.java

```diff
@@ -0,0 +1,310 @@
+    soft.assertThat(result.affectedCommitIds()).contains(c2.id(), e1.id(), e2.id(), e3.id());
+    soft.assertThat(result.affectedCommitIds())
+        .containsAll(b.subList(0, persist.config().parentsPerCommit()));
+    soft.assertThat(result.affectedCommitIds()).hasSize(4 + persist.config().parentsPerCommit());
+  }
+
+  @Test
+  void dryRun() throws Exception {
+    soft.assertThat(repositoryLogic(persist).repositoryExists()).isTrue();
+
+    // Cut point: C1
+    //  root - C1 - C2
+    var root = commit(EMPTY_OBJ_ID, "root");
+    var c1 = commit(root.id(), "c1");
+    var c2 = commit(c1.id(), "c2");
+
+    var cleanup = createCleanup(CleanupParams.builder().dryRun(true).build());
+    CutHistoryParams ctx = cleanup.buildCutHistoryParams(persist, c1.id());
+    CutHistory cutHistory = cleanup.createCutHistory(ctx);
+    CutHistoryScanResult scanResult = cutHistory.identifyAffectedCommits();
+    UpdateParentsResult rewriteResult = cutHistory.rewriteParents(scanResult);
+    UpdateParentsResult cutResult = cutHistory.cutHistory();
+    soft.assertThat(rewriteResult.failures()).isEmpty();
+    soft.assertThat(cutResult.failures()).isEmpty();
+
+    CommitLogic commitLogic = commitLogic(persist);
+    soft.assertThat(commitLogic.fetchCommit(root.id())).isEqualTo(root);
+    soft.assertThat(commitLogic.fetchCommit(c1.id())).isEqualTo(c1);
+    soft.assertThat(commitLogic.fetchCommit(c2.id())).isEqualTo(c2);
+  }
+
+  private IcebergTable table(String key) {
+    return IcebergTable.of(key, 1, 2, 3, 4);
+  }
+
+  private Hash commit(VersionStore store, String ref, Collection<String> keys) throws Exception {
+    ReferenceInfo<CommitMeta> main = store.getNamedRef(ref, GetNamedRefsParams.DEFAULT);
+    CommitResult result =
+        store.commit(
+            BranchName.of(main.getNamedRef().getName()),
+            Optional.of(main.getHash()),
+            CommitMeta.fromMessage("test: " + keys),
+            keys.stream()
+                .map(k -> Operation.Put.of(ContentKey.of(k), table(k)))
+                .collect(Collectors.toList()));
+
+    return result.getCommit().getHash();
+  }
+
+  @ParameterizedTest
+  @ValueSource(ints = {1, 10, 100, 500})
+  void cutPreservesContents(int numExtraKeys) throws Exception {
+    soft.assertThat(repositoryLogic(persist).repositoryExists()).isTrue();
+
+    // Cut point: C1
+    //           /- G3 ---------\
+    //          /- E1 - E2 -\    \ /- F1
+    //  root - C0 --------- C1 - C2 - B1... Bnn
+    //                       \- D1 - D2
+    var main =
+        requireNonNull(repositoryLogic(persist).fetchRepositoryDescription()).defaultBranchName();
+    VersionStore store = new VersionStoreImpl(persist);
+    var root = commit(store, main, List.of("root"));
+
+    Set<String> c0Keys =
+        IntStream.rangeClosed(1, numExtraKeys).mapToObj(i -> "k" + i).collect(Collectors.toSet());
+    var c0 = commit(store, main, c0Keys); // force index stripes to still out
```

Comment:
Better have an assertion that spilling really happened - can be somewhat enforced with random strings in the key

Reply:
added